Whitelisting of resources

simster7 opened the pull request:
As per https://github.com/argoproj/argo-cd/issues/1490, added ability to whitelist resources by including the `resource.inclusions` key in the `argo-cm` ConfigMap. If both resource inclusions and resource exclusions are specified, resource inclusions will take precedence. (See: https://github.com/simster7/argo-cd/blob/f91e7b54c412cfbc9d0c9aee5787f0608e5007a9/util/kube/ctl.go#L84-L92).

Let me know of any desired changes.

alexmt commented:
Awesome, I was hoping we can include it into v1.0.0. Will review today

alexec commented:
An e2e test has failed. You can run those tests locally as follows:

https://argoproj.github.io/argo-cd/developer-guide/test-e2e/

dthomson25 commented:
The E2E test failure was due to a flakey test.  I reran the CI and it passed this time.

simster7 commented:
@alexmt I implemented the changes you requested and added a couple of tests! Let me know what you think.

simster7 commented on util/settings/settings.go:
```diff
@@ -837,21 +861,64 @@ func (a *ArgoCDSettings) GetAppInstanceLabelKey() string {
+// |   Present   |   Present   | Not Allowed |
+// +-------------+-------------+-------------+
+//
+func (a *ArgoCDSettings) IsExcludedResource(apiGroup, kind, cluster string) bool {
```
This is of particular note

alexec commented on util/settings/settings.go:
```diff
@@ -880,14 +880,45 @@ func (a *ArgoCDSettings) checkResourcePresence(apiGroup, kind, cluster string, f
+// +-------------+-------------+-------------+
+//
+func (a *ArgoCDSettings) IsExcludedResource(apiGroup, kind, cluster string) bool {
+	if len(a.ResourceInclusions) > 0 {
```
I feel like this could probably be simpler. 

```golang
return a.isExcludedResource(apiGroup, kind, cluster) || !a.isIncludedResource(apiGroup, kind, cluster)
```

simster7 replied:
I actually thought about this, but I went against this because of a scenario similar to Alex's above:

```
resource.inclusions: |
    - apiGroups:
      - apps
      kinds:
      - DaemonSet
resource.exclusions: |
    - apiGroups:
      - apps
```

When considering a resource with `apiGroups: [apps], kinds: [DaemonSet]`, in the simplified version the operand `a.isExcludedResource(apiGroup, kind, cluster)` would return true for it (it is still under `apiGroups: [apps]`) and the statement would short-circuit and return true. However, if it were allowed to be considered by `a.isIncludedResource(apiGroup, kind, cluster)` it would be allowed and the statement would return false (as expected).

alexec commented on util/settings/filtered_resource.go:
```diff
@@ -5,13 +5,13 @@ import (
 )
 
-type ExcludedResource struct {
+type FilteredResource struct {
```
I feel like you might want to call this `ResourceFilter`?

simster7 replied:
I actually think `FilteredResource` is more appropriate. In this case it reads as a "resource that is filtered" while `ResourceFilter` reads more like "a filter of resources". Since the type/struct refers to a resource and not the filter I think what I have is more appropriate. This is a tad pedantic though, so let me know what you think!

alexec commented on util/settings/settings.go:
```diff
@@ -517,6 +531,16 @@ func (mgr *SettingsManager) SaveSettings(settings *ArgoCDSettings) error {
 		delete(argoCDCM.Data, resourceCustomizationsKey)
 	}
 
+	if len(settings.ResourceInclusions) > 0 {
```
👍  good stuff

alexec commented on util/settings/settings.go:
```diff
@@ -837,21 +861,64 @@ func (a *ArgoCDSettings) GetAppInstanceLabelKey() string {
+	return a.checkResourcePresence(apiGroup, kind, cluster, a.getExcludedResources())
+}
+
+// Behavior of this function is as follows:
```
I think you're describing test cases :). Could you write a [table driven test](https://github.com/golang/go/wiki/TableDrivenTests)?

alexec commented on util/settings/settings_test.go:
```diff
@@ -31,5 +31,91 @@ func TestUpdateSettingsFromConfigMapResourceExclusions(t *testing.T) {
+	assert.Equal(t, []FilteredResource{{APIGroups: []string{}, Kinds: []string{}, Clusters: []string{}}}, settings.ResourceExclusions)
+}
+
+func TestUpdateSettingsFromConfigMapResourceInclusionsAddInclusion(t *testing.T) {
```
Ah! We'll have a merge conflict here as I've made it easier to test this [see code](https://github.com/argoproj/argo-cd/pull/1496/files#diff-c6b0680a5f2fe0ed5aa9197202e013e5R24.). But great work anyway. Now, lets race to see who's PR gets merged first!

simster7 replied:
🏃‍♂️🏃‍♂️🏃‍♂️🏃‍♂️

alexec commented on controller/appcontroller.go:
```diff
@@ -943,6 +944,11 @@ func (ctrl *ApplicationController) watchSettings(ctx context.Context) {
 				prevResourceExclusions = newSettings.ResourceExclusions
 			}
+			if !reflect.DeepEqual(prevResourceInclusions, newSettings.ResourceInclusions) {
+				log.Infof("resource inclusions modified")
```
Mino:but I feel like I'd. want to know what's changed:

```golang
log.WithFields(log.Fields{"prevResourceInclusions": prevResourceInclusions, "newResourceInclusions": newSettings.ResourceInclusions}).Info("resource inclusions modified")
```

simster7 replied:
Added!

simster7 commented:
@alexec I've addressed your comments and conditions! Not sure what you mean by 

> 3. Can you add a test for IsExcludedResource please?

Since I already had tests for `IsExcludedResource` at the end of `settings_test.go`. Could you clarify?

Otherwise this should be ready to go (@alexmt I might need your checkmark on this)
